plugins/inputs/dpdk/dpdk_connector.go

```diff
@@ -0,0 +1,145 @@
+// because otherwise behaviour of connection is undefined (e.g. it could return result of timed out command instead of latest)
+func (conn *dpdkConnector) getCommandResponse(fullCommand string) ([]byte, error) {
+	connection, err := conn.getConnection()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get connection to execute %v command - %v", fullCommand, err)
+	}
+
+	err = conn.setTimeout(connection)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set timeout for %v command - %v", fullCommand, err)
+	}
+
+	_, err = connection.Write([]byte(fullCommand))
+	if err != nil {
+		return nil, conn.tryClose(fmt.Sprintf("send '%v'", fullCommand), err)
+	}
+
+	buf := make([]byte, conn.maxOutputLen)
+	messageLength, err := connection.Read(buf)
+	if err != nil {
+		return nil, conn.tryClose(fmt.Sprintf("read response of '%v'", fullCommand), err)
+	}
+
+	if messageLength == 0 {
+		return nil, fmt.Errorf("got empty response during execution of '%v' command", fullCommand)
+	}
+	return buf[:messageLength], nil
+}
+
+func (conn *dpdkConnector) tryClose(operation string, err error) error {
+	closeErr := conn.connection.Close()
+	conn.connection = nil
+	newErr := fmt.Sprintf("failed to %v command - %v", operation, err)
+	if closeErr != nil {
+		return fmt.Errorf("%v and failed to close connection - %v", newErr, closeErr)
+	}
+	return fmt.Errorf(newErr)
+}
+
+func (conn *dpdkConnector) setTimeout(connection net.Conn) error {
+	if conn.accessTimeout == 0 {
+		return connection.SetDeadline(time.Time{})
+	}
+	return connection.SetDeadline(time.Now().Add(conn.accessTimeout))
+}
+
+// Returns connections, if connection is not created then function tries to recreate it
+func (conn *dpdkConnector) getConnection() (net.Conn, error) {
+	if conn.connection == nil {
+		_, err := conn.connect()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return conn.connection, nil
+}
+
+// Reads InitMessage for connection. Should be read for each connection, otherwise InitMessage is returned as response for first command.
+func (conn *dpdkConnector) readMaxOutputLen(connection net.Conn) (string, error) {
+	buf := make([]byte, maxInitMessageLength)
+	err := conn.setTimeout(connection)
+	if err != nil {
+		return "", fmt.Errorf("failed to set timeout - %v", err)
+	}
+
+	messageLength, err := connection.Read(buf)
+	if err != nil {
+		return "", fmt.Errorf("failed to read InitMessage - %v", err)
+	}
+
+	var initMessage initMessage
+	err = json.Unmarshal(buf[:messageLength], &initMessage)
+	if err != nil {
+		return "", fmt.Errorf("failed to unmarshal response - %v", err)
+	}
+
+	if initMessage.MaxOutputLen == 0 {
+		return "", fmt.Errorf("failed to read maxOutputLen information")
+	}
+
+	if !conn.messageShowed {
+		conn.maxOutputLen = initMessage.MaxOutputLen
+		conn.messageShowed = true
+		return fmt.Sprintf("Successfully connected to %v running as process with PID %v with len %v", initMessage.Version, initMessage.Pid, initMessage.MaxOutputLen), nil
+	}
```

Comment:
I think it would be simpler to return the `initMessage` itself and let the caller decide if and how to print the message.

Reply:
Right, changed.